aptos-move/framework/aptos-framework/sources/vesting.move

```diff
@@ -0,0 +1,1239 @@
+    }
+
+    struct UnlockRewardsEvent has drop, store {
+        admin: address,
+        vesting_contract_address: address,
+        staking_pool_address: address,
+        amount: u64,
+    }
+
+    struct VestEvent has drop, store {
+        admin: address,
+        vesting_contract_address: address,
+        staking_pool_address: address,
+        period_vested: u64,
+        amount: u64,
+    }
+
+    struct DistributeEvent has drop, store {
+        admin: address,
+        vesting_contract_address: address,
+        amount: u64,
+    }
+
+    struct TerminateEvent has drop, store {
+        admin: address,
+        vesting_contract_address: address,
+    }
+
+    struct AdminWithdrawEvent has drop, store {
+        admin: address,
+        vesting_contract_address: address,
+        amount: u64,
+    }
+
+    public fun stake_pool_address(vesting_contract_address: address): address acquires VestingContract {
+        assert_vesting_contract_exists(vesting_contract_address);
+        borrow_global<VestingContract>(vesting_contract_address).staking.pool_address
+    }
+
+    public fun vesting_start_secs(vesting_contract_address: address): u64 acquires VestingContract {
+        assert_vesting_contract_exists(vesting_contract_address);
+        borrow_global<VestingContract>(vesting_contract_address).vesting_schedule.start_timestamp_secs
+    }
+
+    public fun remaining_grant(vesting_contract_address: address): u64 acquires VestingContract {
+        assert_vesting_contract_exists(vesting_contract_address);
+        borrow_global<VestingContract>(vesting_contract_address).remaining_grant
+    }
+
+    public fun beneficiary(vesting_contract_address: address, shareholder: address): address acquires VestingContract {
+        assert_vesting_contract_exists(vesting_contract_address);
+        get_beneficiary(borrow_global<VestingContract>(vesting_contract_address), shareholder)
+    }
+
+    /// Create a vesting schedule with the given schedule of distributions, a vesting start time and period duration.
+    public fun create_vesting_schedule(
+        schedule: vector<FixedPoint32>,
+        start_timestamp_secs: u64,
+        period_duration: u64,
+    ): VestingSchedule {
+        assert!(vector::length(&schedule) > 0, error::invalid_argument(EEMPTY_VESTING_SCHEDULE));
+        assert!(period_duration > 0, error::invalid_argument(EZERO_VESTING_SCHEDULE_PERIOD));
+        assert!(
+            start_timestamp_secs >= timestamp::now_seconds(),
+            error::invalid_argument(EVESTING_START_TOO_SOON),
+        );
+
+        VestingSchedule {
+            schedule,
+            start_timestamp_secs,
+            period_duration,
+            last_vested_period: 0,
+        }
+    }
+
+    /// Create a vesting contract with a given configurations.
+    public fun create_vesting_contract(
+        admin: &signer,
+        grant: Coin<AptosCoin>,
+        shareholders: &vector<address>,
+        shares: &vector<u64>,
```

Comment:
I kinda would prefer vector<(address, Coin<AptosCoin>) and we generate the shares here... but I can do that in my code....

Reply:
So you want to pass a vector<Coin> here as shares (and remove the grant argument)? That works too. How strongly do you feel about it? I can make the change if needed as it's fine either way to me.